Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrType.kt

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+package org.jetbrains.kotlin.native.interop.gen
+
+import org.jetbrains.kotlin.native.interop.gen.jvm.KotlinPlatform
+
+sealed class StubType {
+    abstract val nullable: Boolean
+}
+
+/**
+ * Wrapper over [Classifier].
+ * @property underlyingType is not null if this type is an typealias.
+ */
+class ClassifierStubType(
+        val classifier: Classifier,
+        val typeArguments: List<TypeArgument> = emptyList(),
+        val underlyingType: StubType? = null,
+        override val nullable: Boolean = false
+) : StubType() {
+    fun nested(name: String): ClassifierStubType =
+            ClassifierStubType(classifier.nested(name))
+}
+
+/**
+ * @return type from kotlinx.cinterop package
+ */
+fun KotlinPlatform.getRuntimeType(name: String, nullable: Boolean = false): StubType {
+    val classifier = Classifier.topLevel("kotlinx.cinterop", name)
+    PredefinedTypesHandler.tryExpandPlatformDependentTypealias(classifier, this, nullable)?.let { return it }
+    return ClassifierStubType(classifier, nullable = nullable)
+}
+
+/**
+ * Functional type from kotlin package: ([parameterTypes]) -> [returnType]
+ */
+class FunctionalType(
+      val parameterTypes: List<StubType>,
+      val returnType: StubType,
+      override val nullable: Boolean = false
+) : StubType() {
+    val classifier: Classifier =
+            Classifier.topLevel("kotlin", "Function${parameterTypes.size}")
+}
+
+class TypeParameterType(
+        val name: String,
+        override val nullable: Boolean
+) : StubType()
+
+fun KotlinType.toStubIrType(): StubType = when (this) {
+    is KotlinFunctionType -> this.toStubIrType()
+    is KotlinClassifierType -> this.toStubIrType()
+    else -> error("Unexpected KotlinType: $this")
+}
+
+private fun KotlinFunctionType.toStubIrType(): StubType =
+        FunctionalType(parameterTypes.map { it.toStubIrType() }, returnType.toStubIrType(), nullable)
+
+private fun KotlinClassifierType.toStubIrType(): StubType {
+    PredefinedTypesHandler.tryExpandPredefinedTypealias(classifier, nullable)?.let { return it }
+    val typeArguments = arguments.map { it.toStubIrType() }
+    val underlyingType = underlyingType?.toStubIrType()
+    return ClassifierStubType(classifier, typeArguments, underlyingType, nullable)
+}
+
+private fun KotlinTypeArgument.toStubIrType(): TypeArgument = when (this) {
+    is KotlinType -> TypeArgumentStub(this.toStubIrType())
+    StarProjection -> TypeArgument.StarProjection
+    else -> error("Unexpected KotlinTypeArgument: $this")
+}
+
+/**
+ * Types that come from kotlinx.cinterop require special handling because we
+ * don't have explicit information about their structure.
+ * For example, to be able to produce metadata-based interop library we need to know
+ * that ByteVar is a typealias to ByteVarOf<Byte>.
+ */
+private object PredefinedTypesHandler {
```

Comment:
Doesn't handle `CPointerVar` and maybe other `typealias`es defined in interop runtime.